sprockets/mixins/amqp/__init__.py

```diff
@@ -28,223 +31,481 @@ class PublishingMixin(object):
+                app_id=None, correlation_id=None, properties=None):
+        """Publish a message to RabbitMQ. If the RabbitMQ connection is not
+        established or is blocked, attempt to wait until sending is possible.
+
+        :param str exchange: The exchange to publish the message to.
+        :param str routing_key: The routing key to publish the message with.
+        :param bytes body: The message body to send.
+        :param str app_id: The ID of the app sending the message.
+        :param str correlation_id: The correlation ID associated
+                                   with the message.
+        :param dict properties: An optional dict of additional properties
+                                to append. Will not override mandatory
+                                properties:
+                                app_id, correlation_id, message_id, timestamp
 
-        :param str exchange: The exchange to publish to
-        :param str routing_key: The routing key to publish with
-        :param str message: The message body
-        :param dict properties: The message properties
-        :raises pika.exceptions.ChannelClosed if the channel unexpectedly
-            closes when pika tries to publish.
-        :raises tornado.web.HTTPError if there is a timeout while trying
-            to connect to RabbitMQ. Sends a 504 response.
+        """
+        if self.closing:
+            LOGGER.warning('Connection Closing, waiting for reconnection')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+        elif self.closed:
+            LOGGER.warning('Connection Closed, waiting for reconnection')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+        elif self.blocked:
+            LOGGER.warning('Connection Blocked, waiting for unblock')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+        elif not self.ready:
+            LOGGER.warning('Connection not ready, waiting for ready')
+            yield self._connection_wait(exchange, routing_key,
+                                        body, properties)
+
+        if self.connected:
+            # Set mandatory AMQP properties
+            properties = properties or {}
+            properties['app_id'] = app_id
+            properties['correlation_id'] = correlation_id
+            properties['message_id'] = str(uuid.uuid4())
+            properties['timestamp'] = int(time.time())
+
+            LOGGER.debug('Publishing %d bytes to %s %r (Properties %r)',
+                         len(body), exchange, routing_key, properties)
+            self._channel.basic_publish(
+                exchange,
+                routing_key,
+                body,
+                pika.BasicProperties(**properties),
+                True
```

Comment:
This is a bit opinionated ``mandatory=True`` for my taste, perhaps have it as an argument to the method that defaults to ``False``?